### ISSUE TITLE
Module not found: Can't resolve 'tslib' in '...@aspnet\signalr\dist\esm'

### ISSUE THREAD

**Issue by lghinet** [aspnet/SignalR#2410]
Hello,
i get the following error using "@aspnet/signalr": "1.0.0", and creat-react-app 

```
Failed to compile.

./node_modules/@aspnet/signalr/dist/esm/Errors.js
Module not found: Can't resolve 'tslib' in 'D:\VSTS\LSNG\node_modules\@aspnet\signalr\dist\esm'
```

**Comment by devnixs** [comment]
Just run `npm install --save tslib`

**Comment by anurse** [comment]
It looks like we may have an issue here. `importHelpers` was set but we don't properly depend upon `tslib`. We'll track this for an upcoming release.

**Comment by andreapier** [comment]
Is this issue solved?  
I just updated to v1.0.2 today but am still facing it.

**Comment by anurse** [comment]
It's still open, so no :). You can manually install `tslib` to work around it right now.

**Comment by codepunkt** [comment]
I'm confused. The package, while written in typescript, is published as javascript via npm.

I assume the solution is not to simply add `tslib` as a dependency but fix the build/publish process so that the resulting javascript artifacts don't require `tslib` anymore?

**Comment by BrennanConroy** [comment]
Yeah, which is why the issue is still open and marked as a bug.

**Comment by anurse** [comment]
> I assume the solution is not to simply add tslib as a dependency but fix the build/publish process so that the resulting javascript artifacts don't require tslib anymore?

Yes and no, and we'd actually appreciate some feedback here :). `tslib` is a library containing the helpers that TypeScript would normally generate for things like `await` and class hierarchies. We have two major options here:

1. Change from `importHelpers` to `emitHelpers` in the TypeScript compilation options. This would mean the helpers would be injected into each of the ~40 JS files we have in the `dist/esm` and `dist/cjs` files. That would have an impact on the size of the library, but would avoid the additional dependency
2. Add a proper dependency on `tslib` to our `package.json`. This would fix the actual "bug" since it was the original intent when we enabled `importHelpers`. That would avoid introducing additional code into our compiled JS, but would add the additional dependency. It wouldn't really affect users either way since they use this code (the `tslib` helpers) either way, but it would be a new "dependency" in theory.

I think 2 is our preferred option for a few reasons:

1. Adding the NPM dependency means that users don't see an issue, they just update to our fixed version and the new dependency is installed
2. If a user is using SignalR as well as some other library (such as Angular), they can _share_ the `tslib` dependency, reducing the overall code (note that [`@angular/core`](https://www.npmjs.com/package/@angular/core) also does this, it depends on `tslib`)

So, I think we're going to add the dependency for 2.2.0 and look at back-porting that change to an upcoming 2.1.x patch.

**Comment by stofte** [comment]
I'm a little unsure here, but seems like it might be worth checking out how angular adds their tslib dependence. I'm in an Angular project, and I've not added any references to tslib. So for whatever reasons, Angular doesn't make any demands one or the other in regards to tslib. Seems it would be best for Signalr code to do the same?

**Comment by anurse** [comment]
Yes, this is our issue, as I mentioned above. We're actually removing the `tslib` dependency entirely.

**Comment by eduherminio** [comment]
@anurse Will this fix be included in a future `1.0.3` package version?

**Comment by anurse** [comment]
Yes, we are considering this fix for a patch release.
